reach2

```diff
@@ -1,13 +1,19 @@
 #!/bin/sh
 
-case "$1" in
-  "bootstrap") echo "TODO" ;; # Document this in `optparse` as well
-
-  *)
-    s=$(docker run -it --rm reachsh/reach-cli:latest "$@")
-    if [ "$(echo "$s" | head -n1 | tr -d '\r')" = 'RUNME' ]
-      then echo "$s" | tr -d '\r' | sed 's/RUNME//' | /bin/sh
-      else echo "$s"
-    fi
-    ;;
+IMG='reachsh/reach-cli:latest'
+
+if ! (which make docker docker-compose >/dev/null 2>&1) ; then
+  echo "Reach relies on an installation of make, docker, and docker-compose"
+  exit 1
+fi
+
+if [ "$(docker image ls -q $IMG)" = '' ] ; then
+  docker pull "$IMG"
+  docker run -t --rm "$IMG" update | sed 's/\r//' | sh
+fi
+
+docker run -t --rm "$IMG" "$@" >/tmp/reach.out
+case "$?" in
+  42) sed 's/\r//' /tmp/reach.out | sh ;;
```

Comment:
Why are there `\r`s?

Reply:
This is the Docker output quirk that I alluded to before but I don't have a good answer for why they do that... I just know everything blows up if I don't strip them out before interpreting. My guess without digging too deep is it's something to do with cross-platform concerns like capturing containers' output on Windows.

Running outside of Docker doesn't produce the extraneous `\r`s.

Reply:
What if the script creates a file and then we `docker cp` it out of the container? Maybe this is unnecessary, but I'm scared of arbitrarily changing the output... what if there's `\r` that's supposed to be there?

Reply:
Yeah, perhaps that's better?

I'm actually worried about arbitrary edits too, but every time I consider these edge cases I invariably come back to shell scripts being messy and heavily flawed... I don't know if it's even possible to construct this architecture such that it won't be vulnerable to subtle problems roughly similar to what you're describing (e.g. my hesitation about exit codes, or interpolation being somewhat unavoidable).

I also have a small aversion to leaving (and possibly overwriting) files on the users' filesystem, so I lean towards keeping things in-memory if I can help it.

How would you feel about the wrapper script choosing a "vacant" spot under `/tmp` (rather than overwriting) and cleaning up after itself?

Reply:
I think it is safe to use `/tmp/reach.$$.out`